### PR TITLE
Reflow all bash scripts

### PR DESCRIPTION
**Pull request by wmark**
Converts all bash scripts to bash-programming best practices, standardizes formats where possible, and completes the multi-threaded creation of distributables.

### CODE REVIEW

**Comment by wmark** [comment]
Please hold — don't merge yet.

**Comment by wmark** [comment]
Ready.

https://vine.co/v/iwjHthlamZO

**Comment by mholt** [comment]
Cool, thanks for the tune-up! Will look at it in detail shortly.

File: build.bash [hunk 1]
```diff
@@ -8,48 +8,48 @@
...
 
 pkg=main
+ldflags=()
 
 # Timestamp of build
-builddate_id=$pkg.buildDate
-builddate=`date -u`
+name="${pkg}.buildDate"
+value=$(date --utc +"%F %H:%M:%SZ")
+ldflags+=("-X" "\"${name}=${value}\"")
```
Comment (mholt) [inline build.bash]: But I actually prefer the old/current date format better... can we keep it, pretty plz? :smiley:

**Comment by mholt** [comment]
I guess technically these are both bash scripts; do you think automate.sh should be renamed automate.bash? Now that I'm looking at this... :)

File: build.bash (hunk 1 above)
Comment (wmark) [inline build.bash]: You trolling me here, right? I cannot tell.

There has been no old format. It showed different for every *locale.*

I'd really like not to be the one to re-introduce any particular local format. Follow with a new commit and be verbose about order and rendering of all fields?

----

```sh
$ docker run -ti --rm blitznote/debootstrap-amd64:16.04
# date -u

Sat 2016-03-19 09:51:17 +0000

# vanilla glibc, "C"

Sat Mar 19 09:54:36 UTC 2016

# vanilla glibc, "zh_CN.UTF-8"

2016年 03月 19日 星期六 10:15:56 UTC
```

**Comment by wmark** [comment]
No, we're fine.

The way I see it is: All our *bash* scripts end in ```.sh```; and ```build.bash``` is a hommage to Go's ```build.bash``` and ```make.bash```. Another such hommage would be ```test/runner```.

File: build.bash (hunk 1 above)
Comment (mholt) [inline build.bash]: Alrighty; I'll make the change. Thanks!

**Comment by mholt** [comment]
Running automate.sh on my Mac I get this:

```
./automate.sh: line 68: nproc: command not found
find: -executable: unknown primary or operator
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
```

**Comment by wmark** [comment]
Sorry, the last one is for GNU/Linux.

I could add a bypass for »Darwin«.

File: build.bash [hunk 2]
```diff
@@ -9,9 +9,9 @@
 # Outputs compiled program in current directory.
 # Default file name is 'ecaddy'.
 
-set -e
+set -euo pipefail
 
-: ${output_filename:="$1"}
+: ${output_filename:="${1:-}"}
 : ${output_filename:="ecaddy"}
```
Comment (klingtnet) [inline build.bash]: Why not `${output_filename:="${1:-'ecaddy'}"}`?

**Comment by mholt** [comment]
A bypass would be good, since that's mainly where I run this script. :)

File: build.bash (hunk 2 above)
Comment (wmark) [inline build.bash]: A substitution expression inside the other is frowned upon because it becomes hard to read.

**Comment by mholt** [comment]
@wmark Yes, my Mac has zipnote, apparently.

**Comment by wmark** [comment]
Done. Find the bypass at the bottom of the file.

**Comment by mholt** [comment]
Cool, the bypass worked, but now I get this at about the same time:

```
zipnote error: Bad file descriptor
zipnote error: Temporary file failure (/Users/matt/Dev/src/github.com/mholt/caddy/dist/release/ziGFCapg)
```

**Comment by wmark** [comment]
Can you upgrade zip/zipnote to a patched version? If not, another bypass it is (```mktemp; cp -l; trap rm```).

**Comment by mholt** [comment]
Okay. Well, I really appreciate your work on this. I can't use your automate.sh, but I'll merge this in and revert automate.sh. Thank you!

**Comment by wmark** [comment]
Oh well. Drop me a line when Apple propagated the bugfix for zipnote from 2014 and I will issue the patch to ```dist/automate.sh``` again.

Decreased time spent in packaging stage for me from 11.68s to 1.08s.

**Comment by mholt** [comment]
That's an awesome improvement. I'll keep that in mind!

**Pull request merged**